### ISSUE TITLE
Support fields being required only when publishing, not when saving drafts

### ISSUE THREAD

**Issue by Scotchester** [wagtail/wagtail#12505]
### Is your proposal related to a problem?

In my main Wagtail project, our news pages require the inclusion of a certain field, but the editors need to start drafting the page before the data that goes in that field is available. If the field is set to be required, Django's validation insists that something be in the field, even to save a draft, so editors have to put placeholder data in the field, and then be sure to remember to update that to the real data before publishing.

It would be really great if we had a way to make fields required **only when publishing**, but let drafts be saved without them. 

### Describe the solution you'd like

It's tricky because standard Django model fields don't have any concept of draft vs. published, so we have to work from the premise that the fields are not required at the Django level.

I'm not sure what the best API for handling this would be, but here's one idea: add a new standard attribute called `fields_required_to_publish` to page models, similar to `search_fields`, which accepts a list of strings of the names of fields defined on the model:

```py
class MyPage(Page):
    release_number = models.CharField(blank=True)
    body = RichTextField()
    ...
    fields_required_to_publish = ['release_number']
```

In this simple example, the editor could start drafting a page and filling out its body content without having to specify the release number until publish time.

### Describe alternatives you've considered

There is a [snippet of code floating out there](https://gist.github.com/thibaudcolas/e1cd014ed0c02b46993b1c4a4e134d9b) from a project I'm familiar with that takes the approach of using a decorator on the page model to specify a field that's required to publish. It is currently working for us, but I think the attribute suggestion I gave above feels more in line with established Wagtail patterns.

### Additional context

- The decorator-based alternative includes a `block_schedule` parameter to apply the requirement when scheduling a page for future publishing ("block meaning "prevent" in this case, not referring to StreamField blocks). I personally don't see why you would ever NOT want that to be the case. Setting a future publish time and hitting publish should always presume that it's ready to be published in all other ways except for the timing.
- But speaking of StreamField blocks, it would ALSO be great to have a solution for requiring THOSE only when publishing. Since we control the API for declaring those, we could do that with a simple `required_to_publish` parameter that complements the existing `required` (even for drafts) parameter.
- I hope to follow up soon with some WIP that I started on this at the Wagtail Space US sprint in June.

### Working on this

Anyone can contribute to this. View our [contributing guidelines](https://docs.wagtail.org/en/latest/contributing/index.html), add a comment to the issue once you’re ready to start.

**Comment by laymonage** [comment]
Thanks @Scotchester! I believe this relates to https://github.com/wagtail/rfcs/pull/99, specifically the "Deferring validation until publish" section. Perhaps you could share your thoughts there as well?

> I hope to follow up soon with some WIP that I started on this at the Wagtail Space US sprint in June.

I think that'd be very interesting to see!

**Comment by thibaudcolas** [comment]
👍 I’d be interested to know if there’s ever a case for other types of different validation between drafts and live content. 

And if we did proceed with "deferring validation until publish" – also interesting to consider whether we should then have an API for requiring validation for drafts (🤯).

**Comment by Scotchester** [comment]
> Thanks [@Scotchester](https://github.com/Scotchester)! I believe this relates to [wagtail/rfcs#99](https://github.com/wagtail/rfcs/pull/99), specifically the "Deferring validation until publish" section. Perhaps you could share your thoughts there as well?

Oh, I had seen this and totally forgot about it! The bit about just not validating drafts by default seems like it would solve a lot of the issue:

> To address this, we will relax the validation behaviour so that as standard, validation rules such as blank=False are not enforced when saving drafts. Instead, full validation will be enforced at the point that a page is published, scheduled, or submitted to a workflow.

(It's also a good reminder that there is the more generic issue of validation that isn't just required or not.)

The other thing I'm reminded of is that it would be really nice to have a visible indication that something is fine being blank/invalid for now but _will_ be a problem when you go to publish.

And somewhat dovetailing with all of that… I wish Django itself had a concept of raising a validation _warning_, where it lets you save the form/model but triggers display of a warning message of some kind. Could be useful in lots of scenarios.

**Comment by thibaudcolas** [comment]
> it would be really nice to have a visible indication that something is fine being blank/invalid for now but will be a problem when you go to publish.

Yes, I assume we will display the same messages as we do now, possibly just styled differently. And any saving confirmation would have to indicate it’s "Saved [with errors]" I assume. We definitely don’t want to completely hide validation until publishing.

**Comment by Srishti-j18** [comment]
Hi, can I work on this issue? I'm new to Django and will need to learn a lot to solve it, so it might take some time.

**Comment by gasman** [comment]
@Srishti-j18 Thanks for your interest, but this issue probably won't be a good one for a beginner, as it'll require digging into low-level internal details of both Django and Wagtail.

**Comment by gasman** [comment]
I've come up with a proof-of-concept of this feature at https://github.com/gasman/wagtail/commit/f1ab76d87788a8c52d36b644abe9461b7e3db472, which hopefully aligns with what's required for autosave (RFC 99) - `required` validation rules (including implicit ones from the model field being defined with `blank=False`) are skipped when saving drafts, and only applied on publish / submit. Fields can opt OUT of this behaviour (i.e. revert to the old behaviour of required=True being enforced when saving drafts) by passing `required_on_save=True` on the FieldPanel - this is the default for TitleFieldPanel (so that even draft pages always have a title). I chose to make this an option on FieldPanel as this feels consistent with other options that modify the form's behaviour, such as `permission` and `read_only`.

It works as follows:

* We add a new feature to the base WagtailAdminModelForm: its `Meta` class now accepts a `defer_required_on_fields` attribute, a list of field names. If `form.defer_required_fields()` is called before `form.is_valid()`, these fields will be monkeypatched to unset their `required` attributes. 
* `FieldPanel.get_form_options` is tweaked so that we construct this list at the same time as the usual `fields` and `widgets` attributes.
* `EditView` and `CreateView` for pages have been refactored so that we can find out the requested action before validating the form - if the action is 'save' (as draft), we call `defer_required_fields()` to skip validation.
* We also have to disable model-level validation - the previous behaviour was to call `full_clean` on `Page.save()` and `Page.save_revision()` unless `clean=False` is passed, and this default is now flipped so that the caller has to opt in to cleaning. We'll have to consider this change carefully - some important housekeeping currently happens in `full_clean`, such as assigning a unique slug, and it's possible that developers have added their own custom validation there (although the UX if they do that is somewhat questionable, as it appears we don't catch ValidationErrors raised through there. I guess the assumption is that any validation rules we want to catch and report back to editors in a friendly way should have already been performed at the form level, and the model-level validation is only really for things that bypass the form, such as batch imports...)

**Comment by gasman** [comment]
Completed for Wagtail 7.0 in #12923, #12964, #12984, #13034.

**Comment by janericwinkelmann** [comment]
> [@gasman](https://github.com/gasman) Do we need further work for fields inside StreamField blocks? Thibaud and I tested it the other day and it didn't seem to work ([#13061 (comment)](https://github.com/wagtail/wagtail/discussions/13061#discussioncomment-13014671)), and I seem to recall a similar report on Slack

@gasman Is there any news on this? I was looking into this today for quite a while, because it's pretty undocumented that the new and favorable behavior is indeed NOT working for Blocks inside StreamFields. Because most of our page content is build using StreamFields, this really is a bummer. 

I would love to support on this, but my knowledge around Django and the overall workflow of StreamFields is probably not sufficient enough...

**Comment by thibaudcolas** [comment]
I have created a follow-up feature request for this: [Deferred validation for required fields within StreamField #13699](https://github.com/wagtail/wagtail/issues/13699)
